### DISCUSSION TITLE
[Appello fac-simile] Pipeline

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#81] [Pipeline (M-Z)]
Il programma qui di seguito effettua il conteggio degli elementi pari (in $v0) e dispari (in $v1) presenti nell’array M di lunghezza N.
Si supponga che tutte le istruzioni impiegate fanno parte del set supportato dalla CPU in figura, ossia non si fa uso di alcuna pseudoistruzione.
Si indichino (ignorando hazard che possano concernere la syscall):
$1)$ le istruzioni tra le quali sono presenti **data hazard** – indicare i numeri di istruzione N ed M (visibili alla sinistra delle linee di codice in figura) ed il registro coinvolto $xY (nel formato N / M / $xY);
$2)$ le istruzioni tra le quali sono presenti **control hazard** – indicare i numeri di istruzione N ed M (nel formato N / M);
$3)$ quanti cicli di clock sono necessari ad eseguire il programma tramite **forwarding**, spiegando il calcolo effettuato;
$4)$ quanti cicli di clock sarebbero necessari ad eseguire il programma **senza forwarding**, spiegando il calcolo effettuato;
$5)$ quali sono, per ognuna delle cinque fasi, le **istruzioni** (o le bolle) in pipeline durante **il 14° ciclo di clock** (con **forwarding**);
![Esame](https://github.com/sapienzastudentsnetwork/archelab2223/assets/107944476/26457642-ba0c-49b9-be2a-628cfeb2fec3)
![Architettura Pipeline](https://github.com/sapienzastudentsnetwork/archelab2223/assets/107944476/5ccee98e-7414-49a9-8ea9-856c58d2bed6)

---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665767/AE.-.Appello.fac-simile.-.E4.-.Scheda.di.calcolo.pdf)

**Comment by FeddyLix17** [comment]
## 1) Data Hazards

<details closed>
    <summary> <b> Annotazioni Codice </b> </summary>

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/ec63c138-6ca4-495a-a243-2eadd7b9d849)

</details>

| 9 |  /  |  10  |  /  |  $a1 |
| :---: | :---: | :---: | :---: | :---: |
| 10 |  /  |  11  |  /  | $t7 |
| 10 |  /  |  12  |  /  | $t7 |
| 11 |  /  |  12  |  /  | $t7 |
| 12 |  /  |  13  |  /  | $v0 |
| 15 |  /  |  16  |  /  | $s0 |
| 15 |  /  |  17  |  /  | $s0 |
| 16 |  /  |  17  |  /  | $s0 |
| 21 |  /  | 14  |  /  | $a0 |

## 2) Control Hazards

| 14 |  /  | 23 |
| :---: | :---: |  :---: | 
| 17 |  /  | 20 |

## 3) Cicli di Clock tramite Forwarding

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/4d6e6e9f-19fc-4abb-8c2d-d4ea4ab5a328)

## 4) Cicli di Clock senza Forwarding

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/0fecda20-9214-44fc-bc14-148d76c89df9)

## 5) 14° ciclo di clock (con forwarding)

| Fase | Riga  Codice | Istruzione | 
| :---: | :---: | :---: |
| IF | 18 | addi $v1, $v1, 1 |
| ID | 17 | beq $s0, $0, odd |
| EX | // | bolla |
| MM | 16 | sll $s0, $s0, 31 |
| WB | // | bolla |
